src/lib.rs

```diff
@@ -401,78 +397,79 @@ impl<'hedwig, T, P> PublishBatch<'hedwig, T, P> {
+    /// will not preserve the order.
+    ///
+    /// Users of this function have an opportunity to decide how they want to handle the publishing
+    /// errors that occur. In particular, `PublishBatch` supports retrying failed messages. This
+    /// behaviour can be invoked by returning [`PublishCallbackAction::Retry`] from the closure.
     ///
-    /// Does not consume the builder. Will return `Ok` only if and when all of the messages from
-    /// the builder have been published successfully. In case of failure, unpublished messages will
-    /// remain enqueued in this builder for a subsequent publish call.
-    pub async fn publish(&mut self) -> Result<Vec<P::MessageId>, Error>
+    /// Do note, however, that some kinds of errors may not be transient. An eample of such an
+    /// error is attempting to publish a too large message with the
+    /// [`GooglePubSubPublisher`]. For errors like these retrying is most likely
+    /// incorrect as they would just fail again.
+    ///
+    /// Publisher-specific error types may have methods to make a decision or implementation of the
+    /// callback easier.
+    ///
+    /// [`GooglePubSubPublisher`]: publishers::GooglePubSubPublisher
+    pub async fn publish<F>(&mut self, mut result_callback: F)
     where
         P: Publisher,
+        F: FnMut(&ValidatedMessage, Result<P::MessageId, P::MessageError>) -> PublishCallbackAction,
     {
-        let mut message_ids = Vec::with_capacity(self.messages.len());
-        // Sort the messages by the topic and group them into batches
+        // Sort the messages by the topic…
         self.messages.sort_by_key(|&(k, _)| k);
-        let mut current_topic = "";
-        let mut current_batch = Vec::new();
-        let mut futures_unordered = futures::stream::FuturesUnordered::new();
+        let mut select_all = futures::stream::SelectAll::new();
         let publisher = &self.hedwig.publisher;
-        let make_job = |topic: &'static str, batch: Vec<ValidatedMessage>| async move {
-            (topic, publisher.publish(topic, batch).await)
-        };
-
-        for (topic, message) in mem::replace(&mut self.messages, Vec::new()) {
-            if current_topic != topic && !current_batch.is_empty() {
-                let batch = mem::replace(&mut current_batch, Vec::new());
-                futures_unordered.push(make_job(current_topic, batch));
-            }
-            current_topic = topic;
-            current_batch.push(message)
-        }
-        if !current_batch.is_empty() {
-            futures_unordered.push(make_job(current_topic, current_batch));
+        let mut rest = &self.messages[..];
+        // And group them into batches
+        let mut index = 0;
+        while let Some((topic, _)) = rest.first() {
+            let partition_point = partition_point(rest, |(k, _)| k == topic);
+            let (l, r) = rest.split_at(partition_point);
```

Comment:
given that you don't need indexable subslices per se (just iterators) i think you could replace all this partition/split logic with `itertools::group_by`. The overall index might be the only tricky bit, but I think that should be doable with an `enumerate` on the originating message iterator